Album/ViewControllers/MapControllers/GoogleMapViewController.m

```diff
@@ -123,33 +124,110 @@ - (void)mapView:(GMSMapView *)mapView didLongPressAtCoordinate:(CLLocationCoordi
 }
 
 - (void)mapView:(GMSMapView *)mapView
-    didTapPOIWithPlaceID:(NSString *)placeID
-                    name:(NSString *)name
-                location:(CLLocationCoordinate2D)location {
-  self.infoMarker = [GMSMarker markerWithPosition:location];
-  self.infoMarker.snippet = placeID;
-  self.infoMarker.title = name;
-  self.infoMarker.opacity = 0;
-  CGPoint pos = self.infoMarker.infoWindowAnchor;
-  pos.y = 1;
-  self.infoMarker.infoWindowAnchor = pos;
-  self.infoMarker.map = mapView;
-  mapView.selectedMarker = self.infoMarker;
+didTapPOIWithPlaceID:(NSString *)placeID
+           name:(NSString *)name
+       location:(CLLocationCoordinate2D)location {
+    self.infoMarker = [GMSMarker markerWithPosition:location];
+    self.infoMarker.snippet = placeID;
+    self.infoMarker.title = name;
+    self.infoMarker.opacity = 0;
+    CGPoint pos = self.infoMarker.infoWindowAnchor;
+    pos.y = 1;
+    self.infoMarker.infoWindowAnchor = pos;
+    self.infoMarker.map = mapView;
+    mapView.selectedMarker = self.infoMarker;
 }
 
 - (UIView*) mapView:(GMSMapView *)mapView markerInfoWindow:(nonnull GMSMarker *)marker {
+    // Fetch if there's existing posts related to this coordinate
+    CLLocationCoordinate2D coordinate = marker.position;
+    // If cached data exists (if this coordinate has existing pins)
+    if(self.placeToPins[marker.title]) {
+        InfoMarkerView *markerView = [[[NSBundle mainBundle] loadNibNamed:@"InfoExistWindow" owner:self options:nil] objectAtIndex:0];
+        PFObject *firstPin = [self.placeToPins[marker.title] lastObject];
+        // Set Image
+        NSArray* imagesFromPin = self.pinImages[firstPin.objectId];
+        if(imagesFromPin && imagesFromPin.count > 0) {
+            PFFileObject *imageFile = imagesFromPin[0][@"imageFile"];
+            [markerView.pinImageView setFile:imageFile];
+            [imageFile getDataInBackgroundWithBlock:^(NSData *imageData, NSError *error) {
+                if (!error) {
+                    UIImage *image = [UIImage imageWithData:imageData];
+                    [markerView.pinImageView setImage:image];
+                }
+            }];
+        }
+        // Set place name
+        [markerView.placeNameLabel setText:firstPin[@"placeName"]];
+        // Set date
+        NSString *date = [self.formatter stringFromDate:firstPin[@"traveledOn"]];
+        [markerView.dateLabel setText:date];
+        return markerView;
+    }
+    // Array of pins from the specific coordinate
+    NSArray *pinsFromCoord = [self fetchPinsFromCoord:coordinate];
+    // Check if exisitng pins exist from this coordinate
+    if(pinsFromCoord && pinsFromCoord.count > 0) {
+        PFObject *firstPin = pinsFromCoord[0];
+        int i=0;
+        while(i<pinsFromCoord.count) {
```

Comment:
instead of the while, we could have do something like this:
https://pinkstone.co.uk/how-to-loop-through-every-element-in-an-nsarray/